static/app/views/automations/components/automationBuilderContext.tsx

```diff
@@ -0,0 +1,405 @@
+  _action: AddIfAction
+): AutomationBuilderState {
+  return {
+    ...state,
+    actionFilters: [
+      ...state.actionFilters,
+      {
+        id: state.actionFilters.length.toString(),
+        conditions: [],
+        logicType: DataConditionGroupLogicType.ANY,
+      },
+    ],
+  };
+}
+
+function removeIf(
+  state: AutomationBuilderState,
+  action: RemoveIfAction,
+  formModel?: FormModel
+): AutomationBuilderState {
+  const {groupIndex} = action;
+  if (formModel) {
+    for (const key of formModel.fields.keys()) {
+      if (key.startsWith(`actionFilters.${groupIndex}.`)) {
+        formModel.removeField(key);
+      }
+    }
+  }
+  return {
+    ...state,
+    actionFilters: state.actionFilters.filter((_, i) => i !== groupIndex),
+  };
+}
+
+function addIfCondition(
+  state: AutomationBuilderState,
+  action: AddIfConditionAction
+): AutomationBuilderState {
+  const {groupIndex, conditionType} = action;
+  return {
+    ...state,
+    actionFilters: state.actionFilters.map((group, i) => {
+      if (i !== groupIndex) {
+        return group;
+      }
+      return {
+        ...group,
+        conditions: [
+          ...group.conditions,
+          {
+            comparison_type: conditionType,
+            comparison: {},
+          },
+        ],
+      };
+    }),
+  };
+}
+
+function removeIfCondition(
+  state: AutomationBuilderState,
+  action: RemoveIfConditionAction,
+  formModel?: FormModel
+): AutomationBuilderState {
+  const {groupIndex, conditionIndex} = action;
+  if (formModel) {
+    for (const key of formModel.fields.keys()) {
+      if (key.startsWith(`actionFilters.${groupIndex}.conditions.${conditionIndex}.`)) {
+        formModel.removeField(key);
+      }
+    }
+  }
+  return {
+    ...state,
+    actionFilters: state.actionFilters.map((group, i) => {
+      if (i !== groupIndex) {
+        return group;
+      }
+      return {
+        ...group,
+        conditions: group.conditions.filter((_, j) => j !== conditionIndex),
+      };
+    }),
+  };
+}
+
+function updateIfCondition(
+  state: AutomationBuilderState,
+  action: UpdateIfConditionAction
+): AutomationBuilderState {
+  const {groupIndex, conditionIndex, comparison} = action;
+  return {
+    ...state,
+    actionFilters: state.actionFilters.map((group, i) => {
+      if (i !== groupIndex) {
+        return group;
+      }
+      return {
+        ...group,
+        conditions: group.conditions.map((c, j) =>
+          j === conditionIndex ? {...c, comparison: {...c.comparison, ...comparison}} : c
+        ),
+      };
+    }),
+  };
+}
+
+function updateIfLogicType(
+  state: AutomationBuilderState,
+  action: UpdateIfLogicTypeAction
+): AutomationBuilderState {
+  const {groupIndex, logicType} = action;
+  return {
+    ...state,
+    actionFilters: state.actionFilters.map((group, i) =>
+      i === groupIndex ? {...group, logicType} : group
+    ),
+  };
+}
+
+export function useAutomationBuilderReducer() {
```

Comment:
I'd move this up to the top of the file! 

Generally good to follow what I'd call the "newspaper rule" (probably stolen from someone smarter). "Headline" functions up top, supporting information/utils below.

Reply:
moved in https://github.com/getsentry/sentry/pull/89233/commits/c3b26dc656daeca63eb34f33b18ab0b48f40ce34 !